pkg/allocateip/allocateip.go

```diff
@@ -167,11 +198,36 @@ func ensureHostTunnelAddress(ctx context.Context, c client.Interface, nodename s
 	}
 }
 
-// assignHostTunnelAddr claims an IP address from the first pool
-// with some space. Stores the result in the host's config as its tunnel
-// address. It will assign a VXLAN address if vxlan is true, otherwise an IPIP address.
-func assignHostTunnelAddr(ctx context.Context, c client.Interface, nodename string, cidrs []net.IPNet, vxlan bool) {
-	// Build attributes and handle for this allocation.
+func correctAllocationWithHandle(ctx context.Context, c client.Interface, addr, nodename string, vxlan bool) error {
+	ipAddr := net.ParseIP(addr)
+	if ipAddr == nil {
+		log.Fatalf("Failed to parse node tunnel address '%s'", addr)
+	}
+
+	// Release the old allocation.
+	ipsToRelease := []net.IP{*ipAddr}
+	_, err := c.IPAM().ReleaseIPs(ctx, ipsToRelease)
+	if err != nil {
+		// If we fail to release the old allocation, we shouldn't continue any further. Just exit.
+		log.WithField("IP", ipAddr.String()).WithError(err).Fatal("Error releasing address")
+	}
+
+	// Attempt to re-assign the same address, but with a handle this time.
+	handle, attrs := generateHandleAndAttributes(nodename, vxlan)
+	args := ipam.AssignIPArgs{
+		IP:       *ipAddr,
+		HandleID: &handle,
+		Attrs:    attrs,
+		Hostname: nodename,
+	}
+
+	// If we fail to allocate the same IP, return an error. We'll just
+	// have to allocate a new one. We don't need to update the node object, since
+	// it already has this address attached to it.
+	return c.IPAM().AssignIP(ctx, args)
+}
+
+func generateHandleAndAttributes(nodename string, vxlan bool) (string, map[string]string) {
```

Comment:
Similarly here - I extracted the logic which builds the handle and attributes so it could be resused, since we need it in a couple places now.